data/lua/core/_initial.lua

```diff
@@ -7,14 +7,15 @@ print("Loading Lua core files...")
 local _ = wesnoth.textdomain "wesnoth"
 
 -- Marks a function or subtable as deprecated.
--- Parameters:
----- elem_name: the full name of the element being deprecated (including the module)
----- replacement: the name of the element that will replace it (including the module)
----- level: deprecation level (1-4)
----- version: the version at which the element may be removed (level 2 or 3 only)
----- Set to nil if deprecation level is 1 or 4
----- elem: The actual element being deprecated, ignored if level is 4
----- detail_msg: An optional message to add to the deprecation message
+---@generic T
+---@param elem_name string the full name of the element being deprecated (including the module)
+---@param replacement string the name of the element that will replace it (including the module)
+---@param level '1'|'2'|'3'|'4' deprecation level (1-4)
+---@param version string|nil the version at which the element may be removed (level 2 or 3 only)
+--- Set to nil if deprecation level is 1 or 4
+---@param elem T The actual element being deprecated, ignored if level is 4
+---@param detail_msg? string An optional message to add to the deprecation message
+---@return T
```

Comment:
Here, `T` could do with documentation.

 #5873 has my review of the other stuff here.

Reply:
Does fc5ea0cd372581db57614eaf0b3cad48db84f25c help?

Reply:
> A reference to the element

Something like "a wrapper around" would work better for me. I can understand (and think it's accurate to say) that for functions it returns a wrapper function which prints the warning message.

> --@param elem T The actual element being deprecated, ignored if level is 4

That's true for one use-case of this function, but false and confusing for another, as discussed in #5873.

Reply:
More clarification in 24a0a1055cd8e2e0c35ab0e44958a0463ba53dcd.